src/main/java/ch/njol/skript/registrations/EventValues.java

```diff
@@ -211,118 +211,139 @@ public static <T, E extends Event> T getEventValue(E e, Class<T> c, int time) {
-			if (!c.isAssignableFrom(ev.c))
+		for (EventValueInfo<?, ?> eventValueInfo : eventValues) {
+			if (!c.isAssignableFrom(eventValueInfo.c))
 				continue;
-			if (!checkExcludes(ev, e))
+			if (!checkExcludes(eventValueInfo, event))
 				return null;
-			if (ev.event.isAssignableFrom(e))
-				return (Getter<? extends T, ? super E>) ev.getter;
-			if (!e.isAssignableFrom(ev.event))
+			if (eventValueInfo.event.isAssignableFrom(event))
+				return (Getter<? extends T, ? super E>) eventValueInfo.getter;
+			if (!event.isAssignableFrom(eventValueInfo.event))
 				continue;
 			return new Getter<T, E>() {
 				@Override
 				@Nullable
 				public T get(E event) {
-					if (!ev.event.isInstance(event))
+					if (!eventValueInfo.event.isInstance(event))
 						return null;
-					return ((Getter<? extends T, E>) ev.getter).get(event);
+					return ((Getter<? extends T, E>) eventValueInfo.getter).get(event);
 				}
 			};
 		}
 		// Most checks have returned before this below is called, but Skript will attempt to convert or find an alternative.
 		// Third check is if the returned object matches the class.
-		for (EventValueInfo<?, ?> ev : eventValues) {
-			if (!ev.c.isAssignableFrom(c))
+		for (EventValueInfo<?, ?> eventValueInfo : eventValues) {
+			if (!eventValueInfo.c.isAssignableFrom(c))
 				continue;
-			boolean checkInstanceOf = !ev.event.isAssignableFrom(e);
-			if (checkInstanceOf && !e.isAssignableFrom(ev.event))
+			boolean checkInstanceOf = !eventValueInfo.event.isAssignableFrom(event);
+			if (checkInstanceOf && !event.isAssignableFrom(eventValueInfo.event))
 				continue;
-			if (!checkExcludes(ev, e))
+			if (!checkExcludes(eventValueInfo, event))
 				return null;
 			return new Getter<T, E>() {
 				@Override
 				@Nullable
 				public T get(E event) {
-					if (checkInstanceOf && !ev.event.isInstance(event))
+					if (checkInstanceOf && !eventValueInfo.event.isInstance(event))
 						return null;
-					Object object = ((Getter<? super T, ? super E>) ev.getter).get(event);
+					Object object = ((Getter<? super T, ? super E>) eventValueInfo.getter).get(event);
 					if (c.isInstance(object))
 						return (T) object;
 					return null;
 				}
 			};
 		}
-		// Fourth check will attempt to convert the event value to the type.
-		for (EventValueInfo<?, ?> ev : eventValues) {
-			boolean checkInstanceOf = !ev.event.isAssignableFrom(e);
-			if (checkInstanceOf && !e.isAssignableFrom(ev.event))
+		// Fourth check will attempt to convert the event value to the requesting type.
+		// This first for loop will check that the events are exact. See issue #5016
+		for (EventValueInfo<?, ?> eventValueInfo : eventValues) {
+			if (!event.equals(eventValueInfo.event))
 				continue;
 			
-			Getter<? extends T, ? super E> getter = (Getter<? extends T, ? super E>) getConvertedGetter(ev, c, checkInstanceOf);
+			Getter<? extends T, ? super E> getter = (Getter<? extends T, ? super E>) getConvertedGetter(eventValueInfo, c, true);
```

Comment:
```suggestion
			Getter<? extends T, ? super E> getter = (Getter<? extends T, ? super E>) getConvertedGetter(eventValueInfo, c, false);
```
No need to check instanceof if the events equal

Reply:
The check on line 294 would need to be true, adjusted accordingly.